[Ong Han Yang] iP

FestiveCat opened the pull request:
# ~Duke~ ***Fluffers***
## Added features so far:
- Level-1 to Level-9
- 1000% more [Cats](https://en.wikipedia.org/wiki/Cat) :smile_cat:
> to *meow*, or not to *meow*?
## TODO:
1. A-JUnit
2. Week 4 additions
## **Main method** from `Fluffers::main`
```ruby
    public static void main(String[] args) throws SaveFileModifiedException {
        Fluffers f = new Fluffers();

        System.out.println(f.greet());

        Scanner sc = new Scanner(System.in);

        while (f.isAwake) {
            String input = sc.nextLine();
            System.out.println(f.feedCommandAndReply(input));
        }
    }
```
## Task List
- [x] *Cats*
- [ ] JUnit tests
- [ ] Week 4 tasks

Commit:
Add Gradle support

FestiveCat committed:
Customise Duke into Fluffers, add Echo function

FestiveCat committed:
Add Tasklist, Task, half of lvl2

FestiveCat committed:
Add Task class, remove generic component of TaskList

FestiveCat committed:
add exception for out of index mark/unmark

FestiveCat committed:
OOP-ize Fluffers.java with the input commands, add InvalidInputException

FestiveCat committed:
finish parseInput for deadline, event, todo

FestiveCat committed:
Add simple test case

FestiveCat committed:
add delete method, update regression test

FestiveCat committed:
change ascii art into enum instead of constant

FestiveCat committed:
Add basis for Instruction classes

FestiveCat committed:
finish instructions class

FestiveCat committed:
finish refactoring with instructions class

FestiveCat committed:
update regression tests

FestiveCat committed:
add save feature

FestiveCat committed:
Add date/time support

FestiveCat committed:
Merge branch 'branch-Level-7'

FestiveCat committed:
Merge branch 'branch-Level-8'

FestiveCat committed:
add compatibility between date/time and save/load

FestiveCat committed:
edit gitignore

FestiveCat committed:
add jar

FestiveCat committed:
fix formatting

FestiveCat committed:
fix javadoc style

FestiveCat committed:
fix coding style

FestiveCat committed:
Merge branch 'javadoc'

FestiveCat committed:
Merge branch 'codingstandard'

FestiveCat committed:
add find inst

FestiveCat committed:
make header comments fancy

FestiveCat committed:
Merge branch 'branch-Level-9'

FestiveCat committed:
add checkstyle.xml, fix styles

FestiveCat committed:
Merge branch 'branch-checkstyle'

FestiveCat committed:
Merge branch 'add-gradle-support'

FestiveCat committed:
fix leap year invalid parse

FestiveCat committed:
Merge branch 'date-time-parse-fix'

FestiveCat committed:
fix reset bug

FestiveCat committed:
add JUnit tests

FestiveCat committed:
Merge branch 'A-JUnit'

FestiveCat committed:
Add GUI, Hello Instruction, disable marking/unmarking a task that is already marked/unmarked

FestiveCat committed:
Merge branch 'branch-Level-10'

FestiveCat committed:
make logic clearer in deadline and event instructions

FestiveCat committed:
make logic clearer in file manager

tyanhan commented on src/main/java/chatbots/Chatbot.java:
```diff
@@ -0,0 +1,49 @@
+     * Constructs a chatbot. Initial state of the chatbot is that it is asleep.
+     */
+    protected Chatbot() {
+        this.isAwake = false;
```
It is great that you made your booleans sound like booleans.

tyanhan commented on src/main/java/chatbots/Chatbot.java:
```diff
@@ -0,0 +1,49 @@
+     * @param input the command as input.
+     * @return the reply based on the command.
+     */
+    public abstract String feedCommandAndReply(String input);
```
Would it be better to rename this function to getReply() or something else shorter instead?

FestiveCat replied:
Hmm yeah I think that's a good point, thanks!

tyanhan commented on src/main/java/date/time/DateTimeParser.java:
```diff
@@ -0,0 +1,141 @@
+        // if wrong format, return false
+        // Any "magic numbers" seen from this point onwards refer to the indexes of the specific
+        // string to check, as shown above in line 80.
+        for (int i = 0; i < input.length(); i++) {
```
Would using switch statements help to reduce deep nesting and make your code more efficient?

FestiveCat replied:
Hmmm, what do you mean?
I thought about maybe abstracting each part away into individual helper functions/methods but I'm not too sure how I would use a switch statement here

tyanhan replied:
I was thinking of using switch(i) to check the value of the index instead of multiple if/else statements. This will also help to reduce the amount of nesting and make the code look nicer. But on a second look I guess the amount of nesting wasn't that bad. 

I don't think it is necessary to abstract each individual part as this check is a pretty simple check. Rather than that, 1 suggestion I have(but may not be the most optimal) would be to use String::split and then doing the necessary checks by 1) checking the appropriate lengths of each split, and 2) checking if the splits that are supposed to be digits are digits. The con of this is that it might lengthen your code, but it avoids magic numbers. Let me know what you think.

FestiveCat replied:
Hmm, I'm not particularly fond of using String::split as it relies heavily on the presence of the regex given. This is especially the case where I would need to split multiple times, where I need to check for the length multiple times in addition to creating many objects and eventually enumerating through all the non-separators/dividers to ensure that they are all numbers/digits.

I do agree with how it will remove the magic numbers though, but I think I will keep to this implementation for now? Since I did explain the magic numbers and they aren't too hard to understand

Thank you so much for your input though! I really appreciate it!

tyanhan commented on src/main/java/date/time/DateTimeParser.java:
```diff
@@ -0,0 +1,141 @@
+        }
+        return true;
+    }
+}
```
Perhaps you may want to use LocalDate::parse to help you check for valid date format instead of doing it manually yourself, but at the same time I think it is great that your customised exceptions allow for more intuitive error messages.

tyanhan commented on src/main/java/instructions/Instruction.java:
```diff
@@ -0,0 +1,75 @@
+            return HelloInst.of();
+        case "bye":
+            return ExitInst.of();
+        case "mark":
```
Not sure if it is necessary, but I believe you might have to add //Fallthrough for switch cases without break statement.

FestiveCat replied:
oh good point, I'll be sure to add it, thanks!

tyanhan reviewed (commented):
Great job overall! I did not notice any jarring coding standard errors, and the variable/function names were all well chosen as well. Keep up the good work.

FestiveCat committed:
fix nits from PR feedback

FestiveCat committed:
Refactor DateTimeParser

Refactor DateTimeParser

There are some usages of Magic Numbers used to reference indexes
of the input date-time string, and the main checkValidFormat method
is long (>30LOC).

Breaking the checkValidFormat method down into individual methods that
check for the separators and digits, time and date separately helps to
improve code quality.

Let's split it up so the quality of our code is maintained alongside
the rest of the repo.

FestiveCat committed:
Refactor EventInst and DeadlineInst

Refactor EventInst and DeadlineInst

The factory methods of EventInst and DeadlineInst are messy and hard
to read, especially with all the checks involved.

Abstracting out the checks into methods of their own will help increase
code quality and code readability.

Let's refactor them so they adhere to code quality guidelines.

FestiveCat committed:
add betweentask

FestiveCat committed:
Merge branch 'B-extension'

FestiveCat committed:
add asertions

FestiveCat committed:
Merge branch 'branch-A-Assertions'

FestiveCat committed:
fix code quality

FestiveCat committed:
Merge branch 'branch-A-CodeQuality'

FestiveCat committed:
add better gui

FestiveCat committed:
Update docs/README.md

Replace the original README.md template with one for Fluffers

FestiveCat committed:
Merge pull request #2 from FestiveCat/A-BetterGui

Update docs/README.md and add better gui

FestiveCat committed:
Add usage examples for user guide

FestiveCat committed:
Add line breaks for user guide

FestiveCat committed:
Add description for user guide

FestiveCat committed:
Make .jar release

FestiveCat committed:
Update README.md to indicate ownership of pictures

FestiveCat committed:
Create _config.yml

FestiveCat committed:
Update _config.yml
